### ISSUE TITLE
"model requires more system memory" When Running in Docker Container and Making Continue Plugin Request from Inside Intellij

### ISSUE THREAD

**Issue by nathan-hook** [ollama/ollama#7423]
### What is the issue?

I hope that this a PEBCAK issue and that there is quick environment setting, but with my searching I couldn't find one.

## TL;DR
When using the [Continue Plugin](https://plugins.jetbrains.com/plugin/22707-continue) in my Intellij and then configuring it to talk to my local Docker containerized Ollama instance, I get the following error from the plugin:
```
HTTP 500 Internal Server Error from http://127.0.0.1:11434/api/chat {"error":"model requires more system memory (10.1 GiB) than is available (4.1 GiB)"}
```

## Longer Version
I am running Ollama in a Docker container along with Open WebUI via docker-compose on my Apple M2 Pro with 32 GB or memory:

```
services:
  ollama:
    volumes:
      - ollama:/root/.ollama
    container_name: ollama
    pull_policy: always
    ports:
      - 11434:11434
    environment:
      - OLLAMA_DEBUG=1
    tty: true
    restart: unless-stopped
#    image: ollama/ollama:${OLLAMA_DOCKER_TAG-latest}
    image: ollama/ollama:latest

  open-webui:
    build:
      context: .
      args:
        OLLAMA_BASE_URL: '/ollama'
      dockerfile: Dockerfile
#    image: ghcr.io/open-webui/open-webui:${WEBUI_DOCKER_TAG-main}
    image: ghcr.io/open-webui/open-webui:latest
    container_name: open-webui
    volumes:
      - open-webui:/app/backend/data
    depends_on:
      - ollama
    ports:
#      - ${OPEN_WEBUI_PORT-3000}:8080
      - 3000:8080
    environment:
      - 'OLLAMA_BASE_URL=http://ollama:11434'
#      - 'WEBUI_SECRET_KEY='
    extra_hosts:
      - host.docker.internal:host-gateway
    restart: unless-stopped

volumes:
  ollama: {}
  open-webui: {}
```

In general, these instances have been able to handle most requests that I've made to them.  I have asked general chat questions and then some [Fabric AI](https://github.com/danielmiessler/fabric) queries and Ollama/Open WebUI seemed to do just fine.

When making the requests from Intellij with the Continue plugin, memory suddenly became a problem.

Here is the exact error message from the Continue pluguin:
```
HTTP 500 Internal Server Error from http://127.0.0.1:11434/api/chat {"error":"model requires more system memory (10.1 GiB) than is available (4.1 GiB)"}
```

Then here are the logs from Ollama with the DEBUG=1 environment variable set.  These logs are only show the request from Continue to Ollama:
```
2024-10-30 10:02:05 time=2024-10-30T16:02:05.061Z level=DEBUG source=gpu.go:396 msg="updating system memory data" before.total="23.4 GiB" before.free="4.6 GiB" before.free_swap="6.9 MiB" now.total="23.4 GiB" now.free="4.1 GiB" now.free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.061Z level=DEBUG source=sched.go:181 msg="updating default concurrency" OLLAMA_MAX_LOADED_MODELS=0x7f22c0 gpu_count=1
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=sched.go:211 msg="cpu mode with first model, loading"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=gpu.go:396 msg="updating system memory data" before.total="23.4 GiB" before.free="4.1 GiB" before.free_swap="6.9 MiB" now.total="23.4 GiB" now.free="4.1 GiB" now.free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=INFO source=server.go:105 msg="system memory" total="23.4 GiB" free="4.1 GiB" free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cpu/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v11/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v12/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=memory.go:103 msg=evaluating library=cpu gpu_count=1 available="[4.1 GiB]"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.070Z level=WARN source=server.go:137 msg="model request too large for system" requested="10.1 GiB" available=4443041792 total="23.4 GiB" free="4.1 GiB" swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.070Z level=INFO source=sched.go:428 msg="NewLlamaServer failed" model=/root/.ollama/models/blobs/sha256-ff82381e2bea77d91c1b824c7afb83f6fb73e9f7de9dda631bcdbca564aa5435 error="model requires more system memory (10.1 GiB) than is available (4.1 GiB)"
2024-10-30 10:02:05 [GIN] 2024/10/30 - 16:02:05 | 500 |   18.648125ms |    192.168.65.1 | POST     "/api/chat"
```

Here are my `docker stats`:
```
CONTAINER ID   NAME         CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O        PIDS
2bef887cf4a9   open-webui   0.41%     698.3MiB / 23.44GiB   2.91%     28.4kB / 20.1kB   348MB / 41.5MB   22
29e804e5b9e1   ollama       0.00%     22.76MiB / 23.44GiB   0.09%     9.32kB / 8.49kB   15.2MB / 0B      12
```

My Docker Desktop has the following resource settings:
CPU Limit: 8
Memory Limit: 24 Gigs
SWAP: 4GB

Memory Statistics from my Mac:
Physical Memory: 32 GB
Memory Used: 27.20 GB
Cached Files: 4.75 GB
Swap Used: 2.07 GB

Any friendly direction on how to debug this issue or how to change some environment variables (in docker-compose) to just make this issue go away would be greatly appreciated.

Thank you for all your hard work.

### OS

macOS

### GPU

Other

### CPU

Apple

### Ollama version

ollama version is 0.3.14

**Comment by rick-github** [comment]
If you include more of the log that contains the server environment variables and the calculations that ollama used to assess the memory requirements, the answers would be more certain.  But at a guess, continue is using a large context window that is exacerbated by ollama using a default value of 4 for OLLAMA_NUM_PARALLEL.  These two factors result in a memory requirement to load the model of 10G, which exceeds the available free RAM/swap.  At a base level, the model you are using can fit in just over 5G:
```console
$ ollama ps
NAME                            ID              SIZE    PROCESSOR       UNTIL   
mistral:7b-instruct-v0.3-q4_0   2ae6f6dd7a3d    5.1 GB  100% GPU        Forever
```
which unfortunately is still more than your available resources.  ollama thinks that you have only 7 Megabytes free swap, so despite your config of 4 Gigabytes, something is sucking up those resources.  The upshot is that you don't have enough memory to load this model along with whatever else you are running.  You can try adding more swap, or closing some programs.

**Comment by nathan-hook** [comment]
Thank you for your reply.

The whole logs have been added to the original post and here they are again for reference:
```
2024-10-30 10:01:33 2024/10/30 16:01:33 routes.go:1158: INFO server config env="map[CUDA_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION: HTTPS_PROXY: HTTP_PROXY: NO_PROXY: OLLAMA_DEBUG:true OLLAMA_FLASH_ATTENTION:false OLLAMA_GPU_OVERHEAD:0 OLLAMA_HOST:http://0.0.0.0:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_LLM_LIBRARY: OLLAMA_LOAD_TIMEOUT:5m0s OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_MULTIUSER_CACHE:false OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:0 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://*] OLLAMA_SCHED_SPREAD:false OLLAMA_TMPDIR: ROCR_VISIBLE_DEVICES: http_proxy: https_proxy: no_proxy:]"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.049Z level=INFO source=images.go:754 msg="total blobs: 10"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.050Z level=INFO source=images.go:761 msg="total unused blobs removed: 0"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.050Z level=INFO source=routes.go:1205 msg="Listening on [::]:11434 (version 0.3.14)"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cpu/ollama_llama_server
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v11/ollama_llama_server
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v12/ollama_llama_server
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=INFO source=common.go:49 msg="Dynamic LLM libraries" runners="[cuda_v11 cuda_v12 cpu]"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=DEBUG source=common.go:50 msg="Override detection logic by setting OLLAMA_LLM_LIBRARY"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=DEBUG source=sched.go:105 msg="starting llm scheduler"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.051Z level=INFO source=gpu.go:221 msg="looking for compatible GPUs"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.052Z level=DEBUG source=gpu.go:94 msg="searching for GPU discovery libraries for NVIDIA"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.052Z level=DEBUG source=gpu.go:505 msg="Searching for GPU library" name=libcuda.so*
2024-10-30 10:01:33 time=2024-10-30T16:01:33.052Z level=DEBUG source=gpu.go:528 msg="gpu library search" globs="[/usr/lib/ollama/libcuda.so* /usr/local/nvidia/lib/libcuda.so* /usr/local/nvidia/lib64/libcuda.so* /usr/local/cuda*/targets/*/lib/libcuda.so* /usr/lib/*-linux-gnu/nvidia/current/libcuda.so* /usr/lib/*-linux-gnu/libcuda.so* /usr/lib/wsl/lib/libcuda.so* /usr/lib/wsl/drivers/*/libcuda.so* /opt/cuda/lib*/libcuda.so* /usr/local/cuda/lib*/libcuda.so* /usr/lib*/libcuda.so* /usr/local/lib*/libcuda.so*]"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.053Z level=DEBUG source=gpu.go:562 msg="discovered GPU libraries" paths=[]
2024-10-30 10:01:33 time=2024-10-30T16:01:33.053Z level=DEBUG source=gpu.go:505 msg="Searching for GPU library" name=libcudart.so*
2024-10-30 10:01:33 time=2024-10-30T16:01:33.053Z level=DEBUG source=gpu.go:528 msg="gpu library search" globs="[/usr/lib/ollama/libcudart.so* /usr/local/nvidia/lib/libcudart.so* /usr/local/nvidia/lib64/libcudart.so* /usr/lib/ollama/libcudart.so* /usr/local/cuda/lib64/libcudart.so* /usr/lib/x86_64-linux-gnu/nvidia/current/libcudart.so* /usr/lib/x86_64-linux-gnu/libcudart.so* /usr/lib/wsl/lib/libcudart.so* /usr/lib/wsl/drivers/*/libcudart.so* /opt/cuda/lib64/libcudart.so* /usr/local/cuda*/targets/aarch64-linux/lib/libcudart.so* /usr/lib/aarch64-linux-gnu/nvidia/current/libcudart.so* /usr/lib/aarch64-linux-gnu/libcudart.so* /usr/local/cuda/lib*/libcudart.so* /usr/lib*/libcudart.so* /usr/local/lib*/libcudart.so*]"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.053Z level=DEBUG source=gpu.go:562 msg="discovered GPU libraries" paths="[/usr/lib/ollama/libcudart.so.12.4.99 /usr/lib/ollama/libcudart.so.11.3.109]"
2024-10-30 10:01:33 cudaSetDevice err: 35
2024-10-30 10:01:33 time=2024-10-30T16:01:33.056Z level=DEBUG source=gpu.go:578 msg="Unable to load cudart library /usr/lib/ollama/libcudart.so.12.4.99: your nvidia driver is too old or missing.  If you have a CUDA GPU please upgrade to run ollama"
2024-10-30 10:01:33 cudaSetDevice err: 35
2024-10-30 10:01:33 time=2024-10-30T16:01:33.057Z level=DEBUG source=gpu.go:578 msg="Unable to load cudart library /usr/lib/ollama/libcudart.so.11.3.109: your nvidia driver is too old or missing.  If you have a CUDA GPU please upgrade to run ollama"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.057Z level=DEBUG source=amd_linux.go:416 msg="amdgpu driver not detected /sys/module/amdgpu"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.057Z level=INFO source=gpu.go:384 msg="no compatible GPUs were discovered"
2024-10-30 10:01:33 time=2024-10-30T16:01:33.057Z level=INFO source=types.go:123 msg="inference compute" id=0 library=cpu variant="no vector extensions" compute="" driver=0.0 name="" total="23.4 GiB" available="4.6 GiB"
2024-10-30 10:01:46 [GIN] 2024/10/30 - 16:01:46 | 200 |    3.489458ms |      172.18.0.3 | GET      "/api/tags"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.061Z level=DEBUG source=gpu.go:396 msg="updating system memory data" before.total="23.4 GiB" before.free="4.6 GiB" before.free_swap="6.9 MiB" now.total="23.4 GiB" now.free="4.1 GiB" now.free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.061Z level=DEBUG source=sched.go:181 msg="updating default concurrency" OLLAMA_MAX_LOADED_MODELS=0x7f22c0 gpu_count=1
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=sched.go:211 msg="cpu mode with first model, loading"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=gpu.go:396 msg="updating system memory data" before.total="23.4 GiB" before.free="4.1 GiB" before.free_swap="6.9 MiB" now.total="23.4 GiB" now.free="4.1 GiB" now.free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=INFO source=server.go:105 msg="system memory" total="23.4 GiB" free="4.1 GiB" free_swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cpu/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v11/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=common.go:294 msg="availableServers : found" file=/usr/lib/ollama/runners/cuda_v12/ollama_llama_server
2024-10-30 10:02:05 time=2024-10-30T16:02:05.069Z level=DEBUG source=memory.go:103 msg=evaluating library=cpu gpu_count=1 available="[4.1 GiB]"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.070Z level=WARN source=server.go:137 msg="model request too large for system" requested="10.1 GiB" available=4443041792 total="23.4 GiB" free="4.1 GiB" swap="6.9 MiB"
2024-10-30 10:02:05 time=2024-10-30T16:02:05.070Z level=INFO source=sched.go:428 msg="NewLlamaServer failed" model=/root/.ollama/models/blobs/sha256-ff82381e2bea77d91c1b824c7afb83f6fb73e9f7de9dda631bcdbca564aa5435 error="model requires more system memory (10.1 GiB) than is available (4.1 GiB)"
2024-10-30 10:02:05 [GIN] 2024/10/30 - 16:02:05 | 500 |   18.648125ms |    192.168.65.1 | POST     "/api/chat"
2024-10-30 10:06:17 [GIN] 2024/10/30 - 16:06:17 | 200 |       181.5µs |       127.0.0.1 | GET      "/api/version"
2024-10-30 10:34:21 [GIN] 2024/10/30 - 16:34:21 | 200 |      232.25µs |       127.0.0.1 | HEAD     "/"
2024-10-30 10:34:21 [GIN] 2024/10/30 - 16:34:21 | 200 |     281.833µs |       127.0.0.1 | GET      "/api/ps"
2024-10-30 10:47:05 [GIN] 2024/10/30 - 16:47:05 | 200 |    1.833166ms |      172.18.0.3 | GET      "/api/tags"
2024-10-30 10:47:05 [GIN] 2024/10/30 - 16:47:05 | 200 |      77.917µs |      172.18.0.3 | GET      "/api/version"
2024-10-30 10:47:16 [GIN] 2024/10/30 - 16:47:16 | 200 |      61.084µs |      172.18.0.3 | GET      "/api/version"
2024-10-30 10:47:22 [GIN] 2024/10/30 - 16:47:22 | 200 |     223.334µs |      172.18.0.3 | GET      "/api/version"
```

I will try closing everything down and trying again.

Thanks for the advice.

Besides shutting everything down, are there any other nobs that could be turned or switches that could be flipped to get things to work?

**Comment by rick-github** [comment]
```
2024-10-30 12:33:33 time=2024-10-30T18:33:33.774Z level=INFO source=server.go:105 msg="system memory" total="23.4 GiB" free="21.5 GiB" free_swap="3.9 GiB"
```
At the time the model was being loaded, there was 21.5G RAM and 3.9G swap, plenty of room for the 10.3G required for the model:
```
2024-10-30 12:33:33 time=2024-10-30T18:33:33.775Z level=INFO source=memory.go:326 msg="offload to cpu" layers.requested=-1 layers.model=33 layers.offload=0 layers.split="" memory.available="[21.5 GiB]" memory.gpu_overhead="0 B" memory.required.full="10.3 GiB" memory.required.partial="0 B" memory.required.kv="4.0 GiB" memory.required.allocations="[10.3 GiB]" memory.weights.total="7.6 GiB" memory.weights.repeating="7.2 GiB" memory.weights.nonrepeating="411.0 MiB" memory.graph.full="2.1 GiB" memory.graph.partial="2.2 GiB"
```
Since `OLLAMA_NUM_PARALLEL` is unset, ollama uses a default value of 4.  Combined with a context size of 8K set by Continue, the total context size allocated is 32K (`--ctx-size`).  This contributes to the size of the KV cache allocated, 4G (`memory.required.kv`, above).  This is the allocation that varies depending on the number of parallel queries and the size of the context window.  Setting `OLLAMA_NUM_PARALLEL=1` will reduce this allocation to around 1G.
```
2024-10-30 12:33:33 time=2024-10-30T18:33:33.777Z level=INFO source=server.go:388 msg="starting llama server" cmd="/usr/lib/ollama/runners/cpu/ollama_llama_server --model /root/.ollama/models/blobs/sha256-8eeb52dfb3bb9aefdf9d1ef24b3bdbcfbe82238798c4b918278320b6fcef18fe --ctx-size 32384 --batch-size 512 --embedding --verbose --threads 8 --no-mmap --parallel 4 --port 37131"
```
My guess for the most likely culprit as the memory hog earlier would be the browser.   Extensions and pages that present lots of visual effects are notorious for sucking up RAM.

**Comment by corrosive4354** [comment]
If possible, try to increase the memory limit of Docker to a higher value, such as 28 GiB or higher. Make sure to adjust it in the settings of Docker Desktop.

In fact, Docker is not very compatible with Mac.
You can choose[ Servbay](https://servbay.com) to try, which is more professional than docker on Mac

**Comment by remon-nashid** [comment]
Same here. Issue started occurring with ollama v0.3.14, even without docker or continue dev.

**Comment by nathan-hook** [comment]
Thank you for closing this ticket and sorry for not replying before now...

Closing the browser and reducing used ram was the culprit.  Once I closed my browser and only had a window with a couple of tabs open everything started to work as expected.

I have a terrible habit of keeping way too many browser tabs/windows open.

Thank you again for your help and insights.
